datapath: add tail call hooks for custom metrics, bytecounter example

qmonnet opened the pull request:
## Custom Metrics

Byte counters! Let's count how much data is received and sent for each endpoint. Related PR: #13173

But more than byte counters, this PR introduces hooks in the datapath so that users can tail call into custom programs, and collect all kind of metrics they want - Provided they implement it themselves in eBPF. Here is an overview:

- First commit introduces the tail call hooks in the datapath (IPv4 only), plus management for the per-endpoint prog_array maps used to reference the custom programs. Those maps have two entries, one for ingress and one for egress.

- Second commit makes these tail calls optional, and opt-in, through the introduction of a dedicated Cilium option.

- Third commit adds a sample application: a simple byte counter which can be attached to keep track of the amount of data received or sent for a given endpoint.

- Fourth commit adds the tail call hooks to the IPv6 datapath, on the same model as for IPv4.

To use it, once Cilium is deployed:

- Compile the custom program
  ```bash
  make -C bpf/custom
  ```
- Load it
  ```bash
  bpftool prog load bpf/custom/bpf_custom.o /sys/fs/bpf/tc/globals/bytecounter type classifier \
          pinmaps /sys/fs/bpf/tc/globals/bytecounter_maps
  ```
- Reference it from the prog_array map
  ```bash
  EP=<endpoint_id>
  bpftool map update pinned /sys/fs/bpf/tc/globals/cilium_calls_custom_$(printf '%05d' ${EP}) \
          key 0 0 0 0 value pinned /sys/fs/bpf/tc/globals/bytecounter              # or key `1 0 0 0` for egress
  ```
- For now, dump counters' value manually
  ```bash
  bpftool map dump pinned /sys/fs/bpf/tc/globals/bytecounter_maps/bytecount_map
  ```

## PR Status

Here are some points of interest:

- <s>I am still trying to finalise the location for the ingress hook. In the current version, the tail call added to `tail_ipv4_policy()` is not always executed, as `handle_policy()` (which calls it) is skipped in `ipv4_local_delivery()` if `USE_BPF_PROG_FOR_INGRESS_POLICY` is defined i.e. if `enable-endpoint-routes` is set to `true`. This is the case for GKE, for example. So my current plan is to add the tail call to the different branches in `ipv4_local_delivery()`, possibly by replacing `return`s with `goto` in that function to write the tail call just once, and also possibly with preprocessor guards to make sure it is only called from bpf_lxc.</s>

- The custom program needs to return with the same value the original program would have used, to preserve the datapath logics. This means that we must pass the desired value from bpf_lxc to the custom program we hook into. I'm using `cb[CB_CT_STATE]` for that because it did not seem to be used at that location, but I might be wrong. Feedback on this approach (or an alternatives) would be welcome.

- <s>I also pass the current direction (ingress or egress) and the source identity to the custom program. Given that direction needs 1 bit, identity needs 24, and existing return values are low and needs less than 7 bits, I make it all fit on the same `cb` cell.</s>

- Once we have the counters in an eBPF map, we would need to extract and process it. It could be a Cilium metric, but I'm not sure how to implement custom, user-defined metrics. It could be just from the command line, but similarly this requires to give a way to users to specify what kind of metrics they want. At the moment, the values from the counters can be retrieved from bpftool.

qmonnet committed:
datapath: add tail call hooks to run programs for custom metrics

Add tail call hooks to bpf_lxc in order to run custom programs. The
objective is to attach user-specified programs in order to collect
custom metrics.

Three tail calls are added, one on the egress path, the other two for
distinct branches on the ingress path. The custom program is passed
through dedicated, per-endpoint prog_array maps. These maps are created
by tc when loading bpf_lxc programs, and managed by the Cilium agent
afterwards. Each prog_array map has two entries, one for ingress and one
for egress. Make sure we skip the tail call when the packet is being
redirected to a L7 proxy, to avoid running the custom program twice on
the incoming packet.

In this commit, tail call is attempted unconditionally. Follow-up commit
will guard it with a macro so it can be removed at compilation time.

On failure to jump through the tail call, the behaviour of the bpf_lxc
program is mostly unchanged. On success, it jumps to a custom program
and never goes back to the original program. The hook is inserted at the
very end of the original program, to avoid skipping packet processing
operations. However, to preserve a consistent behaviour, the return
value from the final eBPF programs must remain the same. To that end,
the return value is passed through the cb buffer so the custom program
can eventually return it.

At this stage, custom programs have to be loaded, attached, and
referenced in the prog_array maps by the user (typically with clang and
bpftool).

Custom tail calls are not supported for host endpoints at this stage.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
config: add opt-in option for tail call hooks for custom metrics

Add a new option --enable-custom-calls to Cilium, in order to select
whether tail calls for custom metrics (specified as a user-defined
program) should be compiled and called in the eBPF datapath.

The option defaults to "false", meaning that the tail call hooks are
compiled out by default.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
datapath: add sample program for custom metrics (byte counter)

Add a sample program implementing custom metrics using the dedicated
tail call hooks in the datapath. The result is a per-endpoint byte
counter for ingress and egress.

This comes as a new subdirectory under bpf/. It contains four files:

- bpf_custom.c is considered as the program entry point and as the
  "landing point" for the tail call hook. It ensures that the return
  value selected in the datapath, and stored in the cb buffer, is
  effectively returned. It also calls a "custom_prog()" function, which
  is to be implemented by the user, and passes it the context and the
  identity for the packet. The content in bpf_custom.c is _not_ expected
  to be edited by the user. The #include directive is used to specify
  the header file containing the core of the user program. The name of
  the file to include is passed via a macro BPF_CUSTOM_PROG_FILE, which
  is typically set by the Makefile.
  This file is dual-licensed (GPL-2.0-only / BSD-2-Clause) to allow for
  more flexibility on user-defined programs.

- bytecount.h is supposed to be written by a user willing to implement
  custom metrics collection. It collects data and possibly stores it in
  a map, as the user sees fit. On exit it returns to bpf_custom.c, where
  the final return value for the eBPF program is picked according to the
  datapath decision. This user-defined program is _not_ expected to tail
  call out of bpf_custom.c, or to modify the packet, at the risk of
  breaking the normal datapath behaviour.

- A Makefile is provided to help compile those programs. They are
  automatically built on a normal invocation of "make -C bpf". One
  advantage of building the custom program with this Makefile is that
  the other eBPF libraries from Cilium (under bpf/lib/) are available.
  This Makefile is largely inspired from the one under bpf/sockops, some
  consolidation might be conceivable in the future.
  It is possible to specify the file to include for loading the custom
  function to use in the eBPF program, when calling make. This is done
  with "BPF_CUSTOM_PROG_FILE=custom_file.h make -C bpf". If not set, the
  value defaults to bytecount.h, provided in this commit. Similarly, the
  variable BPF_CUSTOM_PROG_NAME helps configuring the name of the ELF
  section where the program is stored, to change the name or the program
  type (depending on how the loader interprets this information).

- A README.rst provides instructions on the step that are required to
  use the example byte counter.

CI-wise, note that because we use a macro for the section name,
verifier-test.sh cannot retrieve the section name correctly. We cheat
and have it parse the comment with the default value instead.

Resolves: #13173
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno commented on bpf/lib/maps.h:
```diff
@@ -99,7 +99,7 @@ struct bpf_elf_map __section_maps TUNNEL_MAP = {
 #endif
 
-#if (defined CUSTOM_CALLS_MAP)
+#if (defined ENABLE_CUSTOM_CALLS) && (defined CUSTOM_CALLS_MAP)
```
Nit: Why the parenthesis?

qmonnet replied:
They are misplaced, they should start before the macro names and not before the `if`s. I'll fix it, thanks.

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -543,6 +547,9 @@ func (h *HeaderfileWriter) writeStaticData(fw io.Writer, e datapath.EndpointConf
 		callsMapName = callsmap.HostMapName
 	}
 	fmt.Fprintf(fw, "#define CALLS_MAP %s\n", bpf.LocalMapName(callsMapName, epID))
+	if option.Config.EnableCustomCalls {
```
Can we also avoid defining `CUSTOM_CALLS_MAP` for the host endpoint (`e.IsHost()`)?

qmonnet replied:
Makes sense, I'll look into it.

pchaigno commented on bpf/custom/bpf_custom.c:
```diff
@@ -0,0 +1,29 @@
+int custom_hook(struct __ctx_buff *ctx)
+{
+	int ret = ctx_load_meta(ctx, CB_CUSTOM_CALLS);
+	int identity = get_identity(ctx);
```
This retrieves the identity from the packet mark. AFAIU, we're not guaranteed to have encoded the identity there. I'm unsure what the impact is here.

If this is specific to the current byte counter use case, maybe we should move this into `custom_prog()`? If we want to always provide the security identity to the custom program (i.e., users can assume it's part of the API), maybe we could pass the identity via metadata instead?

qmonnet replied:
I believe it will be helpful to collect metrics and I would like to have it part of the API somehow. I did not realise the identity might not always be encoded, thanks. I'll explore ways to pass it through the `cb` buffer, although one concern is to find an offset in `cb` which is unused at this stage. I'm using one for passing the datapath return value already (although we only use a few possible return values I think, so maybe I could encode several values on the 32bits of one `cb` cell if I don't come up with a better solution).

pchaigno commented on bpf/custom/bytecount.h:
```diff
@@ -0,0 +1,24 @@
+};
+
+static __always_inline
+void custom_prog(struct __ctx_buff *ctx, __u32 identity)
```
Since the packet isn't supposed to be modified by the custom program, could be mark `ctx` as `const`?

qmonnet replied:
No, this would prevent us from using eBPF helpers taking a pointer to the context, even if they're read-only.

pchaigno replied:
Hm. Can't we fix the function prototype for those helpers?

qmonnet replied:
Maybe. I'll bring this up at tomorrow's datapath meeting I think.

pchaigno commented on bpf/custom/bpf_custom.c:
```diff
@@ -0,0 +1,29 @@
+	int identity = get_identity(ctx);
+
+	/* Call user-defined function from custom header file. */
+	custom_prog(ctx, identity);
```
Do we want to use a more specific name for this hook? We may need to consider this name part of the API because changing it will require (small) fixes for users of this new feature. If we later add new custom hooks in the datapath, we will need to add new functions similar to this one (and can't call them all `custom_prog` :slightly_smiling_face: ).

qmonnet replied:
`custom_prog` has the advantage of being quite explicit for the review :) But agreed, we can probably find something else before we merge. A simple alternative would be to rename `bpf/custom/bpf_custom.c` into `bpf/custom_metrics/bpf_custom_metrics.c`, and `custom_prog()` function into something like `collect_metrics()`? Or do you have something else in mind?

pchaigno replied:
I think we should clarify the position of the hook in the datapath somehow because its function (e.g., collect metrics) really depends on what users will do with it. So maybe something like `post_container_policy`?

qmonnet replied:
But then if we had other hooks elsewhere in the future, we could theoretically reuse the same `bpf_custom.c` as a landing point, but then the name would not apply anymore? (Thinking of it, it also applies to something with `custom_metrics` in its name). Hmm :thinking:

pchaigno replied:
> But then if we had other hooks elsewhere in the future, we could theoretically reuse the same bpf_custom.c as a landing point

That was actually my initial concern. I don't think we'll want to reuse that function for other (new) hooks because otherwise existing custom programs will start receiving a lot more events (i.e., packets) than they expect. If the program is counting packets for example, it may all of a sudden start counting packets twice or start counting packets it doesn't want to count.

qmonnet replied:
I'm not sure I'm following you. You can reuse the `bpf_custom.c` base, but to produce a different programme that would be attached to another hook. If the user-defined part is different, they should be no risk of having packets counted multiple times?

pchaigno reviewed (commented):
Thanks for all the comments and explanations! That makes it very easy to review :-)

A few nits below, I haven't commented on the position of hooks within `bpf_lxc` since we discussed that during #sig-datapath meeting already.

My main remaining concern is around the API we expose. I'm assuming we won't be able to change what we expose to custom program after the fact because it may break custom programs in the wild. I've added a couple comments below in that regard.

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -543,6 +547,9 @@ func (h *HeaderfileWriter) writeStaticData(fw io.Writer, e datapath.EndpointConf
 		callsMapName = callsmap.HostMapName
 	}
 	fmt.Fprintf(fw, "#define CALLS_MAP %s\n", bpf.LocalMapName(callsMapName, epID))
+	if option.Config.EnableCustomCalls && !e.IsHost() {
```
I think we also need to skip some logic in case of host endpoint during the object file patching. Otherwise we'll get warnings. I believe the CI tests should catch this.

joestringer replied:
Specifically I'd expect this:

```
scopedLog.Warning("Skipping symbol substitution")
```

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -50,6 +50,20 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
+
+	custom_meta += (identity & 0xffffff);
+	custom_meta += (ret & 0xff) << 24;
+	custom_meta += (direction & 0x1) << 31;
```
Nit: I find the use of `+=` a bit misleading here, maybe `|=`?

joestringer replied:
Also, `0x80000000` is a little ambiguous as it can be set by either the `ret` or the `direction`. Although in practice I think only low positive integer values are likely to be passed in `ret` for the forseable future, it probably makes sense to have the `ret` use a mask of `0xEF`?

qmonnet replied:
Agreed for the `|=`, and the `0xff` instead of `0xef` is totally a mistake on my side. I'll fix it, thanks.

aanm reviewed (approved):
Reviewed the agent side changes and they seem pretty small to me

joestringer commented on bpf/lib/maps.h:
```diff
@@ -99,6 +99,25 @@ struct bpf_elf_map __section_maps TUNNEL_MAP = {
+	.size_key	= sizeof(__u32),
+	.size_value	= sizeof(__u32),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 2,	/* ingress and egress */
```
The custom program seems to encode directionality in the `CB`, but this map hints at two different programs for ingress/egress. Which path do we want to support?

qmonnet replied:
Right, so there two ways to dissociate ingress/egress path:
- By attaching different programs in the two entries of the endpoint map.
- By encoding the direction and passing it down to the custom program.

The PR indeed supports both. I started with the different entries in the endpoint map, which has the advantage of skipping the tail call entirely for one direction if we only want to attach a program to the other direction. But I ended up attaching identical programs to collect data on both directions except the data would be stored in a different eBPF map entry based on the direction, and I figured that making the direction accessible to the custom program would be great to avoid compiling two otherwise identical programs.

With hindsight, I understand that mixing the two approaches may look confusing. I could remove the encoding of the direction in the `cb` buffer.

joestringer commented on bpf/custom/bpf_custom.c:
```diff
@@ -0,0 +1,34 @@
+	int ret = (custom_meta >> 24) & 0xff;
+
+	/* Call user-defined function from custom header file. */
+	custom_prog(ctx, identity, direction);
```
If the custom program called certain helper functions (I'm thinking like `bpf_redirect()`), then it would influence the path that the `skb` will take. How much guardrails do we want to put in place for cases like this?

pchaigno replied:
During the last #sig-datapath meeting, we discussed using `const` (including in helpers) to have at least a compile-time warning against that.

qmonnet replied:
I don't see an easy way to prevent such scenario. We indeed discussed making the `ctx` pointer `const` to discourage modifications of the packet. But ultimately, we have no guarantee that the users will use the `bpf_custom.c` “landing point” for the custom hook, they can as well build and load their own programs outside of the Cilium repo.

I don't think we should look for ways to strongly prevent that, but instead this should be correctly documented (TBD). We should make very clear that modifying the packet or the return value at this stage has a huge chance of crashing the datapath. But if we can make sure that people are aware of that and they prefer to experiment and live on the edge, I think we should not make it a fight to load custom-custom programs.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -50,6 +50,20 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
+#ifdef ENABLE_CUSTOM_CALLS
+static __always_inline void
+encode_custom_prog_meta(struct __ctx_buff *ctx, __u32 ret, int direction,
+			int identity)
```
nit: The types used here seem a little inconsistent, `ret` is typically a signed `int` in the parent functions while `identity` is typically a `__u32`.

joestringer replied:
I was also wondering whether there's anything we should do about `ret` being both unsigned and > 8bits, to prevent possible mixups if someone unintentionally passed large or negative values in here.

qmonnet replied:
(I knew I wanted to double-check types somewhere before pushing!)
Sure, will fix. That's right, making sure the return value is between the expected bounds make sense, it's probably much safer to skip the tail call entirely if this is not the case. We won't be able to return the expected value otherwise.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -50,6 +50,20 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
+#ifdef ENABLE_CUSTOM_CALLS
+static __always_inline void
+encode_custom_prog_meta(struct __ctx_buff *ctx, __u32 ret, int direction,
+			int identity)
```
nit: `identity` seems to reflect the "local" endpoint (eg source for `direction == CUSTOM_CALLS_IDX_INGRESS` and destination for `direction == CUSTOM_CALLS_IDX_EGRESS`). Perhaps this should be indicated somewhere, either via the variable name or a comment?

qmonnet replied:
True, some polishing required on that side. I'll address it.

joestringer commented on bpf/custom/bpf_custom.c:
```diff
@@ -0,0 +1,34 @@
+#include "lib/common.h"
+
+#define DIRECTION_INGRESS	0
+#define DIRECTION_EGRESS	1
```
nit: Do we need these defined here as well as `CUSTOM_CALLS_IDX_INGRESS`, `CUSTOM_CALLS_IDX_EGRESS` since the caller is using the latter to pass through? Or if we think it's nicer to read, maybe we should define these against the aforementioned constants?

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -756,6 +770,11 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
 
+#ifdef ENABLE_CUSTOM_CALLS
+	encode_custom_prog_meta(ctx, ret, dstID, CUSTOM_CALLS_IDX_EGRESS);
+	tail_call_static(ctx, &CUSTOM_CALLS_MAP, CUSTOM_CALLS_IDX_EGRESS);
```
Any thoughts on IPv6 support?

qmonnet replied:
I plan to add it, and I believe it will be mostly similar to IPv4, with the hooks to be added to the same points in the datapath. For that reason I have only focused on IPv4 so far, to make it easier to review and address feedback. I intend to add it as a new commit next time I push to this PR.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1243,6 +1262,12 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
+#ifdef ENABLE_CUSTOM_CALLS
+	encode_custom_prog_meta(ctx, ret, src_label, CUSTOM_CALLS_IDX_INGRESS);
+	tail_call_static(ctx, &CUSTOM_CALLS_MAP, CUSTOM_CALLS_IDX_INGRESS);
+#endif
```
I note that this seems to run on egress prior to reaching the egress proxy, and on ingress prior to reaching the ingress proxy. This doesn't seem entirely balanced. Do you have a model in mind of how this should interact with L7 proxies?

qmonnet replied:
The placement of the tail call hooks was mostly dictated by the constraints coming with the hooks: I run the custom program at the very last stage, as I cannot “jump back” to the datapath code.

I probably didn't give as much attention to the L7 proxy as I should have, so no, I don't really have a model in mind at the moment.

Reading through the code, it looks like the decision to redirect to the proxy is taken in `ipv4_policy()`, but is not achieved before we return from `tail_ipv4_policy()`, _after_ the tail call hook. So I'm wondering... Maybe I'm actually running the custom program on each trip to and from the proxy? I need to double-check and run tests. If that's the case I can add a restriction on `ret == POLICY_ACT_PROXY_REDIRECT` I suppose. Maybe I can find a way to run _after_ L7 proxy on ingress to get something more balanced?

What information exactly would we loose if the hooks remain unbalanced?

qmonnet replied:
I did have a double processing, before and after the packet was processed by the ingress proxy. Addressed with a guard on `ret != CTX_ACT_TX`.

joestringer reviewed (changes requested):
This looks pretty clean to me, nice work.

The main questions that come to mind reviewing this PR are:
* Custom metrics on packets that are subject to L7 forwarding seems a little inconsistent
* Seems like there are two approaches for handling ingress/egress hinted in the PR, either having two programs (one ingress / one egress) or passing the direction bit in the `CB`. Is this just left over from a previous implementation or did you intend to support more than one custom program?
* Some helpers could be called by a custom program and influence the handling of the packet, such as `bpf_redirect()`. What kind of guard rails do we want for cases like this? (One simple answer can be just leave it up to the implementer not to mess up; or if we're more concerned we could put compile-time warnings in place to discourage such behaviour)
* Is IPv6 support intended? How will we handle this?

I've added comments on each of these topics below so we can have threaded discussions.

Beyond that, there's just a scattering of minor nits, none of which are overly consequential.

qmonnet committed:
datapath: add tail call hooks for custom programs to IPv6 datapath

On the same model that the hooks added to the IPv4 datapath, add hooks
for tail calls to user-defined eBPF programs to provide a way to
implement, for example, custom metrics collection for IPv6 flows.

The prog_array maps holding the references to the custom programs are
extended with two new cells, for IPv6 ingress and egress paths.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -53,6 +53,32 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
+ * executing tail calls to custom programs. "ret" is the return value supposed
+ * to be returned to the kernel, needed by the callee to preserve the datapath
+ * logics. The "identity" reflects the local endpoint: the source of the packet
+ * on ingress path, or its destination on the egress path. We encode it so that
```
nit, on first read I found it ambiguous whether the "reflects" + "source of the packet" were trying to imply that the "identity" parameter is really an endpoint ID or a security identity that is common across multiple endpoints. A couple of extra words help to dispell that:

```suggestion
 * logics. The "identity" is the security identity of the local endpoint: the source of the packet
 * on ingress path, or its destination on the egress path. We encode it so that
```

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -774,6 +798,10 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
 
+	if (!encode_custom_prog_meta(ctx, ret, dstID))
+		tail_call_static(ctx, &CUSTOM_CALLS_MAP,
+				 CUSTOM_CALLS_IDX_IPV4_EGRESS);
```
Should we still increment some metric if the tail call fails?

qmonnet replied:
Why not, that would make sense. Do you think I should `DROP_MISSED_TAIL_CALL`, or add new dedicated metrics?

qmonnet replied:
[Unresolving] So I went from `DROP_MISSED_TAIL_CALL`, but I'm reverting this now, this is not what we want. We do expect a lot of those tail calls to not happen, because the hook is active as soon as we enable the support for custom programs in the agent but it will effectively be taken only if the user has set up a custom program to be called in the prog_array map.

So if we want to know how many times the tail call “fails” (or should we say, does not occur), we would need another metric. But I'm not sure how relevant this would be, because you could have “missed” tail calls just in the duration between the LXC program is attached, and the prog_array map for the endpoint is updated from the user, and this should not necessarily be seen as a bug or anything.

joestringer replied:
If we assume that the custom progs are dynamically added & removed at any point during runtime then yeah I'm not sure what use it will be to track these. I can imagine coming up with some solution like dynamically enabling such a metric programmatically from Cilium via static data + reload or something but I feel like that's overly complicated and the benefit seems low.

At any given point in time, we can dump the custom tail calls prog array map and see whether the prog is there and infer whether it's being called or not. (Which reminds me, we should add this optional map dump to the debuginfo...)

pchaigno replied:
A metric is not an error or a warning. It's up to the user to interpret them. We already report a lot of metrics (e.g., packet drops, regenerating endpoints, etc.) which can, under the right circumstances, be interpreted as either benign counters or errors. In addition, I'd expect most users to collect metrics over time and watch their evolution rather than their absolute value.

Right now, if a user suspects the tail call is missed, they have no clear way to validate. They have to implement some sort of logging in their own program and trust it to be correct. Not the most pain-free way to start extending Cilium's datapath IMO.

qmonnet replied:
You have a point. Do you mind if I open a follow-up issue once this is merged so we can come back on this? I'd like to avoid adding cycles on implementation, naming, and testing of the metrics in this PR.

One note on metrics evolution: In this case, the metrics will likely always be increasing when hooks for custom programs are enabled, unless _all_ hooks have a destination program attached (or are unused in the datapath), so this may be harder to interpret. But as you say, it's probably better than no logging.

pchaigno replied:
> Do you mind if I open a follow-up issue once this is merged so we can come back on this?

Sure :+1: Let's just :crossed_fingers: that it doesn't introduce a complexity issue.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1293,6 +1321,11 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	/* Store meta: essential for proxy ingress, see bpf_host.c */
 	ctx_store_meta(ctx, CB_PROXY_MAGIC, ctx->mark);
+
+	if (ret != CTX_ACT_TX && !encode_custom_prog_meta(ctx, ret, src_label))
```
Maybe a comment would be useful here to explain the first part of the condition is for the proxy case?

qmonnet replied:
Agreed, I was careful to document in the commit log, but forgot the comment. Will fix.

pchaigno commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -71,6 +71,7 @@ cilium-agent [flags]
       --enable-bpf-clock-probe                               Enable BPF clock source probing for more efficient tick retrieval
```
Small typo in this commit message (2nd commit): `in order to select weather tail calls`.

qmonnet replied:
It's a thing here, the weather jumps from sunny to rainy in an instant. Stack is preserved and all. Famous weather tail calls.

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -476,6 +476,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	}
 
+	if option.Config.EnableCustomCalls {
+		cDefinesMap["ENABLE_CUSTOM_CALLS"] = "1"
```
Could you also add this macro to `bpf/Makefile`, for the complexity tests and to be compile tested?

pchaigno reviewed (commented):
LGTM. Only a couple minor comments below. I haven't tested it.

Do we need to test it in CI? How could we do that?

aditighag reviewed (commented):
LGTM! Follow-up comments (some of them discussed offline) so that others can also pitch in - 
- It'll be good to add a CI test that validates this code. ICMP traffic might be more deterministic to validate test success criteria.
- If possible, replace the `bpftool` driven workflow with `cilium/ebpf` library in the new CI test.
- Document unsupported packet processing paths like socket LB.

joestringer reviewed (changes requested):
I didn't go through with a comb again yet but one nit below + general feedback we already discussed around:
* What should testing look like for this feature, which you've already discussed briefly above, and
* Understanding where we want to get to for the "loading" side of this. This looks currently usable by advanced users with some extra scripting / YAML-foo around the edges but if we intend to tidy this up more, it'd be nice to see what we think the loading pieces should look like and how that influences the user workflow.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -53,6 +53,32 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
+	 */
+	if ((ret & 0xff) != ret)
+		return -1;
+	custom_meta |= (ret & 0xff) << 24;
+	custom_meta |= (identity & 0xffffff);
```
Is this specifically trying to mirror `set_identity_mark()` in terms of bit layout? Would it make sense to either make a reference or refactor to reuse that code specifically to ensure this? (I realize this is putting the result into `skb->cb[]` instead of `skb->mark` though).

qmonnet replied:
It does _not_ attempt to mirror that layout. From what I understand, setting/retrieving the identity from and to the `skb->mark` needs this specific setup because 8 of the 24 bits were added [as an afterthought](https://github.com/cilium/cilium/pull/4858/files), and could not be placed next to the other 16 ones in the field.

No such constraint here: Just take the full 24 bits at once.
Before that we use the `ret` return value, because we need it in the custom program, but this is unrelated to what `set_identity_mark()` does.

So I do not believe there is much to refactor here.

joestringer replied:
Ah yes, I saw the pattern and assumed it was the same but now I look again, it's clearly just putting the bits all together.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1036,6 +1077,21 @@ int tail_ipv6_policy(struct __ctx_buff *ctx)
+	 * incoming packet (before redirecting, and on the way back from the
+	 * proxy).
+	 */
+	if (ret != CTX_ACT_TX &&
```
I think that the `ret != CTX_ACT_TX` is making an assumption about the way we forward to the proxy here when really we should just be checking the condition that occurs on line 1072. At that earlier point in the function, `ret == POLICY_ACT_PROXY_REDIRECT` means that we are definitely sending the traffic to the proxy.

The assumption is that to forward to the proxy, a device redirect is required. That is not the case if `from_host` is true in this context. The `CTX_ACT_TX` here means we will redirect to another device to continue processing, but if for instance we are performing a policy lookup at source device for host->pod on same node, then I don't think that we will set the `ret` to `CTX_ACT_TX` in that case when forwarding traffic to the proxy. So we'll double-trace that particular case.

(Presumably the same for IPv4).

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1096,6 +1152,15 @@ int tail_ipv6_to_endpoint(struct __ctx_buff *ctx)
 					ret, CTX_ACT_DROP, METRIC_INGRESS);
+
+#ifdef ENABLE_CUSTOM_CALLS
+	if (!encode_custom_prog_meta(ctx, ret, src_identity)) {
```
This code doesn't seem to handle the proxy case, but we know it could occur up on line 1150 with the hairpin (which calls redirect and passes back the return code).

joestringer replied:
In contrast to the case above, I think for this case it's probably OK to use `CTX_ACT_TX` because it's in the "to_endpoint" variation which is only ever run on the TC egress qdisc on the way into a pod so we *must* redirect the packets to send them through the proxy. If we didn't actually redirect the packets here then we would just let the packets through to the pod without traversing the proxy. That said, for consistency maybe it'd be better to make the logic consistent between the two cases.

qmonnet replied:
Thank you, and agreed for the consistency. I'll update.

joestringer reviewed (changes requested):
I've added a few more notes below from a deeper look.

aditighag commented on bpf/custom/bpf_custom.c:
```diff
@@ -0,0 +1,30 @@
+ */
+#include "bytecount.h"
+
+__section("custom")
```
Encode the program type (`classifier/custom`) in the section name so that `cilium/ebpf` loader can infer the type?

qmonnet replied:
I have not done so at this time, but might be open to update it in the future. I understand it would make it easier to load the program from the Go library. On the other hand, it forces us to set a specific program type, and we loose the context flexibility TC/XDP. This probably needs some more reflection...

aditighag replied:
Agreed. The loader does have another API where you can specify the program type while loading a BPF program, but I'm still figuring out how to get the asm instructions in a more direct way.

joestringer replied:
We could always templatize the section name as well, something like (immediately above this line):

```
#ifndef BPF_CUSTOM_PROG_NAME
#define BPF_CUSTOM_PROG_NAME "custom"
#endif
```

plus

```suggestion
__section(BPF_CUSTOM_PROG_NAME)
```

qmonnet committed:
helm: add option for enabling tail calls to custom programs in datapath

Transcribe the --enable-custom-calls Cilium option to the Helm charts.

The documentation was then updated with:

    make -C install/kubernetes docs

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test: add CI test for tail calls hooks for custom programs

Add a new CI test to check that custom programs loaded and run via the
datapath tail call hooks work as expected. In particular, we use the
byte counter example provided in a previous commit.

The test is the following: We deploy Cilium and some pods, we load the
byte counter program and attach it to a hook (ingress or egress) for a
given endpoint, we send some traffic (ping requests and replies), and
then check the values in the eBPF map where the count is stored. Try
again with per-endpoint routes, to check the remaining ingress hook.

A new manifest is added, to make sure that we have two simple
application endpoints, located on different nodes. This is to check that
the egress hook is working (there is no egress hook for socket-based
load-balancing yet).

This manifest also pulls an image which is used to compile the byte
counter program before we can load it. This image should be updated in
the future to embed the sample byte counter and all related headers, so
we wouldn't have to mount a volume against the source repository on the
node. This would permit to remove the byte counter from Cilium's
repository, and to keep it in that dedicated image instead.

Skip the test on 4.9 kernels because bpftool map updates do not work on
such kernels. Skip on GKE because we do not have Cilium's source to
compile the custom program from. Both issues should be addressed in the
future, by moving the program sources to a dedicated image which should
also embed its own loader. But actually, just run on net-next, since the
coverage for other kernel versions would be the same.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,146 @@
+       # rm /sys/fs/bpf/tc/globals/custom_prog
+       # rm -r /sys/fs/bpf/tc/globals/bytecounter_maps
+
+Using Other Custom Programs
+===========================
```
nit: Are these instructions substantially different than the above? At a glance they seem very similar and I figure anyone writing additional custom statistics programs would likely just modify this custom example to get where they're going.

Not a big deal, I'd personally probably just skip the instructions below but if you think they should stay then feel free to just mark this comment as resolved.

qmonnet replied:
Makes sense. I kept the section but simplified it and removed all the duplicated instructions (essentially keeping the `make` command with the environment variables to set a different name for the file to include).

qmonnet committed:
test/bpf/verifier-test.sh: improve ELF section names collection

Improve ELF section names collection in preparation for adding a sample
custom program for the datapath to the list of the programs to test. The
function had the two following limitations:

- Grepping on "__section(" would attempt to extract a section name for
  all lines containing this pattern, but the sed expression supposed to
  extract the name needs an opening double quote inside the parenthesis.
  In the case when the argument is not a string (but a macro, for
  example) and this double quote is missing, the sed command fails, and
  the whole line is returned as a section name. Address this by ignoring
  the lines without the opening double quote.

- It would not discard the beginning of the line, and would consider
  anything written _before_ symbol "__section" as part of the section
  name. Address this by discarding anything before "__section".

While at it, remove the call to "grep", since we use sed already and
it's perfectly able to do the filtering itself.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on test/k8sT/CustomCalls.go:
```diff
@@ -0,0 +1,306 @@
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to lookup byte counter value ('%s'): %s", cmd, res.Stderr()))
+
+			// Output from bpftool is on the form
+			// "0x11\n0x22\n0x33\n...0x88\n", convert it.
+			// Too bad bpftool returns hex that jq doesn't know how
+			// to convert yet.
```
(Completely unrelated to this PR) Definitely feels like there's a gap in the overall JSON workflow here if you can't just pipe bpftool into jq to do the operations you want on the data. Maybe BTF can solve this for us some day.

qmonnet replied:
My fault, actually :( I wasn't much experienced with JSON when I added it to bpftool, and I had the terrible idea of making it print hexa strings. There is a built-in function in jq (`tonumber`) to parse strings as integers but it doesn't support hex strings yet (I found an open issue for that).

I'm thinking of adding an option (a new one, or maybe double `-jj`|`-pp`) to have bpftool dump real integers. Feels dirty, but I guess hex strings are UAPI by kernel's standards now :(.

joestringer replied:
Given that the values are arbitrary length and often make the most sense as hexadecimal if they can't otherwise be better-interpreted, I can understand avoiding encoding directly as JSON integer types since the common display format would likely render them in decimal. Is bpftool output format really subject to kernel UAPI stability though?

qmonnet replied:
That was my understanding. Maybe the maintainers would be more flexible than for traditional kernel UAPI, but it's hard to say. Changing it would likely break some scripts. That would be a discussion for bpf@vger I suppose.

joestringer reviewed (approved):
Assuming this passes CI, if we don't get any further responses from @cilium/ci-structure then I'll be inclined to just go ahead and merge this soon.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1348,7 +1352,9 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	bool from_host = ctx_load_meta(ctx, CB_FROM_HOST);
+#ifdef ENABLE_CUSTOM_CALLS
 	bool proxy_redirect = false;
+#endif
```
Can we just mark it as `__maybe_unused`?

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1359,7 +1365,9 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 		ret = ctx_redirect_to_proxy4(ctx, &tuple, proxy_port, from_host);
+#ifdef ENABLE_CUSTOM_CALLS
 		proxy_redirect = true;
+#endif
```
These `#ifdef`s are very noisy. I think we can just drop them since the compiler should be able to optimize out `proxy_redirect`.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1385,7 +1395,9 @@ declare_tailcall_if(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)),
 	__u32 src_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
+#ifdef ENABLE_CUSTOM_CALLS
 	bool proxy_redirect = false;
+#endif
```
Ditto.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1433,13 +1445,16 @@ int tail_ipv4_to_endpoint(struct __ctx_buff *ctx)
 		ret = ctx_redirect_to_proxy_hairpin(ctx, proxy_port);
+#ifdef ENABLE_CUSTOM_CALLS
 		proxy_redirect = true;
+#endif
```
Ditto.

pchaigno commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -358,6 +358,11 @@ containerRuntime:
+# -- Tail call hooks for custom eBPF programs.
+customCalls:
+  # -- Enable tail call hooks for custom eBPF programs.
+  enable: false
```
Shouldn't this be `enabled`?

We also need to have a discussion at some point about `feature=true` vs. `feature.enabled=true`. Right now we're mixing both :S

joestringer replied:
in general, it should be `feature.enabled=true` otherwise there's no way to extend the configuration of the parent object.

pchaigno commented on test/bpf/verifier-test.sh:
```diff
@@ -61,7 +61,7 @@ function cleanup {
 
 function get_section {
-	grep "__section(" $DIR/$1 | sed 's/__sec[^\"]*\"\([0-9A-Za-z_-]*\).*/\1/'
+	sed -n '/__section("/ s/.*__section("\([0-9A-Za-z_-]*\).*/\1/p' $DIR/$1
```
> - It would collect the argument passed to __section(), even if this
  argument is not a string (but a macro, for example) and does not
  reflect the final value passed to the compiler. Address this by
  requiring an opening double quote in the pattern we use for detection.

I don't understand that part of the commit message. The double quote was already required by the sed.

I actually think it causes us to skip the templated section in `bpf_lxc`: https://github.com/cilium/cilium/blob/9d94081cef70ea8a05b32d9c81baf64ec8d6f316/bpf/bpf_lxc.c#L1412 (We don't need to fix that in this PR of course.)

qmonnet replied:
The sed requires the double quote. The grep does not. So for a `__section()` with a macro name, the whole line is taken unchanged, see:

```
$ echo '__section("foo.h")'    | grep "__section(" | sed 's/__sec[^\"]*\"\([0-9A-Za-z_-]*\).*/\1/'
foo
$ echo '__section(SOME_MACRO)' | grep "__section(" | sed 's/__sec[^\"]*\"\([0-9A-Za-z_-]*\).*/\1/'
__section(SOME_MACRO)
$ echo '__section(SOME_MACRO)' | sed -n '/__section("/ s/.*__section("\([0-9A-Za-z_-]*\).*/\1/p'
$
```

My comment is incorrect in the sense that if there's a macro instead of a string, the whole line (and not the argument passed to `__section()`) would be collected as a section name.

The `__section_tail()` is left aside by the `grep` pattern as well. The presence of the double quote in the sed expression would make the command fail to extract the correct name, not skip it.

I'll update the commit log to make it more accurate and easier to understand. I think I can also remove the `/__section("/` pattern match at the beginning of my own expression, and keep only the substitute command. This should have the same effect, but we would have just one pattern to update if we want to support the `__section_tail()` in a follow-up.

pchaigno commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,125 @@
+
+.. note::
+    This is a beta feature. Please provide feedback and file a GitHub issue if
+    you experience any problems.
```
FYI, I think this syntax is specific to Read the Docs. GitHub won't render it: https://gist.github.com/pchaigno/9397bdad9553204f93fee0226d384f1e.

qmonnet replied:
Right, I suspected so but forgot to check, thank you. FYI it is part of the directives [documented](https://docutils.sourceforge.io/docs/ref/rst/directives.html#attention) and implemented by docutils, which also defines the specifications for RST, but indeed not rendered on GitHub. I'll fix.

pchaigno commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,125 @@
+--------------
+
+When compiled with ``ENABLE_CUSTOM_CALLS`` set, Cilium's datapath contains some
+tail call hooks permitting to insert custom programs in the datapath. These
```
```suggestion
 tail call hooks to insert custom programs in the datapath. These
```

pchaigno commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,125 @@
+  custom program must be called as the very last step of the processing of the
+  packet, and must preserve the return value selected by the datapath.
+
+- The identity corresponding to the packet, that is, the Cilium identity
```
```suggestion
The remote security identity associated with the packet, that is, the Cilium identity
```

pchaigno commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,125 @@
+- The identity corresponding to the packet, that is, the Cilium identity
+  associated to the sender for the ingress hook, and to the receiver when on
+  the egress hook. This identity is not strictly necessary to the custom
+  program, but it prove useful for some use cases. Given that the datapath
```
```suggestion
  program, but it proved useful for some use cases. Given that the datapath
```

pchaigno commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,125 @@
+Limitations
+-----------
+
+- There is currently no hook on the egress path for socket-based
```
```suggestion
 - There is currently no hook on the egress path for socket-level
```

pchaigno commented on bpf/custom/README.rst:
```diff
@@ -0,0 +1,125 @@
+Sample Byte Counter Program
+============================
+
+This directory contains a sample byte counter program, to help count how many
```
```suggestion
This directory contains a sample byte-counter program, to help count how many
```

qmonnet replied:
Updated elsewhere in this doc and CI test.

pchaigno commented on test/k8sT/CustomCalls.go:
```diff
@@ -0,0 +1,306 @@
+
+			// Deploy Cilium, enable tail calls to custom programs
+			deploymentManager.DeployCilium(map[string]string{
+				"customCalls.enabled": "true",
```
Could we run the same with per-endpoint routes as well? Otherwise, one of the three IPv4 hook is not covered.

qmonnet replied:
Good point, thank you!

pchaigno commented on test/k8sT/CustomCalls.go:
```diff
@@ -0,0 +1,306 @@
+		// Also skip on GKE because we do not have the source of the
+		// custom program available on a node, for the compiler pod to
+		// pick up.
+		return helpers.DoesNotRunOnNetNextOr419Kernel() ||
```
Do we need to run on both net-next and 4.19? Will these two runs cover different things?

qmonnet replied:
No, I don't expect the coverage would change. I proceeded by eliminating incompatible setups rather than picking the ones strictly needed. I can restrict to e.g. net-next if we think this is enough.

qmonnet replied:
Switched to net-next only.

pchaigno commented on test/k8sT/CustomCalls.go:
```diff
@@ -0,0 +1,306 @@
+	})
+
+	AfterFailed(func() {
+		kubectl.CiliumReport("cilium status", "cilium endpoint list")
```
Would other bpftool outputs be useful here? E.g., maps, progs content of new tail call map.

qmonnet replied:
Possibly. I'll look at this, and at the potential bugtool additions, as a follow-up.

pchaigno reviewed (commented):
Mostly minor comments and nits below. The two things that may need to be addressed here instead of a follow up are (1) a typo in `values.yaml` and (2) the test which should be run with per-endpoint routes as well.

pchaigno commented on test/k8sT/CustomCalls.go:
```diff
@@ -0,0 +1,362 @@
+			identityKey := getIdentityKey("k8s:id=app1", ciliumPodK8s2)
+
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
```
There's a lot of code duplication between the two `It` blocks. I'm wondering if we could move most of the above code to a `BeforeAll` or `BeforeEach`. I'd keep the `DeployCilium` in the `It` block though.

Is it possible to install the custom program before Cilium is deployed?

qmonnet replied:
Currently the program is copied to the Cilium pod and loaded from there, so this sounds unlikely. But it may be enough to load it on the node itself, in which case this could indeed be simplified in the future.

I tried to factor as much as I could, but `copyAndLoadObjectFile()`, `getPodsInfo()`, `kubectl.GetCiliumEndpoint()`, `getIdentityKey()` all need Cilium to be deployed at the moment. One option could be to have a function that takes the options for deploying Cilium, and the two `It()` blocks simply calling it with different parameters.

pchaigno replied:
> One option could be to have a function that takes the options for deploying Cilium, and the two `It()` blocks simply calling it with different parameters.

:+1:

pchaigno reviewed (approved):
LGTM!

One comment below, but probably best to address/discuss as a follow up.
